roachtest: add admission/follower-overload

tbg opened the pull request:
This is a less ad-hoc version of the experiment in #81289, where I
messed with the EBS configuration. This can't be done programmatically,
and so here we use an IO nemesis on n3 instead.

```
open "$(roachprod adminui $c:1)/#/debug/chart?charts=%5B%7B%22metrics%22%3A%5B%7B%22downsampler%22%3A3%2C%22aggregator%22%3A3%2C%22derivative%22%3A0%2C%22perNode%22%3Afalse%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.store.raft.quota_pool.percent_used-max%22%7D%2C%7B%22downsampler%22%3A3%2C%22aggregator%22%3A3%2C%22derivative%22%3A0%2C%22perNode%22%3Afalse%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.store.raft.quota_pool.percent_used-p50%22%7D%5D%2C%22axisUnits%22%3A0%7D%2C%7B%22metrics%22%3A%5B%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A0%2C%22perNode%22%3Atrue%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.store.storage.l0-sublevels%22%7D%5D%2C%22axisUnits%22%3A0%7D%2C%7B%22metrics%22%3A%5B%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A2%2C%22perNode%22%3Atrue%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.node.sys.host.disk.write.bytes%22%7D%5D%2C%22axisUnits%22%3A1%7D%2C%7B%22metrics%22%3A%5B%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A0%2C%22perNode%22%3Afalse%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.node.exec.latency-p50%22%7D%2C%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A0%2C%22perNode%22%3Afalse%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.node.exec.latency-p99%22%7D%5D%2C%22axisUnits%22%3A2%7D%2C%7B%22metrics%22%3A%5B%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A2%2C%22perNode%22%3Afalse%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.node.exec.latency-count%22%7D%5D%2C%22axisUnits%22%3A0%7D%2C%7B%22metrics%22%3A%5B%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A0%2C%22perNode%22%3Atrue%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.store.storage.l0-num-files%22%7D%5D%2C%22axisUnits%22%3A0%7D%2C%7B%22metrics%22%3A%5B%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A2%2C%22perNode%22%3Atrue%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.store.rocksdb.compacted-bytes-written%22%7D%5D%2C%22axisUnits%22%3A1%7D%2C%7B%22metrics%22%3A%5B%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A0%2C%22perNode%22%3Atrue%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.store.replicas.leaseholders%22%7D%5D%2C%22axisUnits%22%3A0%7D%2C%7B%22metrics%22%3A%5B%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A2%2C%22perNode%22%3Atrue%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.node.admission.wait_durations.kv-count%22%7D%5D%2C%22axisUnits%22%3A0%7D%2C%7B%22metrics%22%3A%5B%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A0%2C%22perNode%22%3Atrue%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.node.sys.cpu.combined.percent-normalized%22%7D%5D%2C%22axisUnits%22%3A0%7D%2C%7B%22metrics%22%3A%5B%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A0%2C%22perNode%22%3Atrue%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.node.sys.go.totalbytes%22%7D%5D%2C%22axisUnits%22%3A1%7D%2C%7B%22metrics%22%3A%5B%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A2%2C%22perNode%22%3Atrue%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.store.raft.rcvd.admission_wait%22%7D%5D%2C%22axisUnits%22%3A2%7D%2C%7B%22metrics%22%3A%5B%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A2%2C%22perNode%22%3Atrue%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.store.raft.rcvd.dropped_bytes%22%7D%5D%2C%22axisUnits%22%3A1%7D%2C%7B%22metrics%22%3A%5B%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A0%2C%22perNode%22%3Atrue%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.store.raft.rcvd.queued_bytes%22%7D%5D%2C%22axisUnits%22%3A1%7D%2C%7B%22metrics%22%3A%5B%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A0%2C%22perNode%22%3Atrue%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.store.raft.process.handleready.latency-p75%22%7D%2C%7B%22downsampler%22%3A1%2C%22aggregator%22%3A2%2C%22derivative%22%3A0%2C%22perNode%22%3Atrue%2C%22source%22%3A%22%22%2C%22metric%22%3A%22cr.store.raft.process.handleready.latency-p50%22%7D%5D%2C%22axisUnits%22%3A2%7D%5D"

open "http://$(roachprod ip --external $c:4):9090/graph?g0.expr=storage_l0_sublevels&g0.tab=0&g0.stacked=0&g0.range_input=1h&g1.expr=storage_l0_num_files&g1.tab=0&g1.stacked=0&g1.range_input=1h&g2.expr=rate(rocksdb_compacted_bytes_written%5B1m%5D)&g2.tab=0&g2.stacked=0&g2.range_input=1h&g3.expr=histogram_quantile(0.90%2C%20rate(workload_kv_write_duration_seconds_bucket%5B1m%5D))&g3.tab=0&g3.stacked=0&g3.range_input=1h&g4.expr=histogram_quantile(0.90%2C%20rate(workload_kv_read_duration_seconds_bucket%5B1m%5D))&g4.tab=0&g4.stacked=0&g4.range_input=1h&g5.expr=rate(workload_kv_write_duration_seconds_count%5B1m%5D)&g5.tab=0&g5.stacked=0&g5.range_input=1h&g6.expr=histogram_quantile(1.0%2C%20rate(raft_quota_pool_percent_used_bucket%5B1m%5D))&g6.tab=0&g6.stacked=0&g6.range_input=1h&g7.expr=histogram_quantile(1.0%2C%20rate(admission_wait_durations_kv_stores_bucket%5B1m%5D))&g7.tab=0&g7.stacked=0&g7.range_input=1h"
```

Part of https://github.com/cockroachdb/cockroach/issues/79215.
Closes #81834.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/81516)
<!-- Reviewable:end -->

tbg commented:
Here's what this roachtest looks like in action (10min) https://www.loom.com/share/f6bec2f6ce26449f9bb04e269375b048

tbg committed:
roachtest: add admission/follower-overload

This is is a less ad-hoc version of the experiment in #81289, where I
messed with the EBS configuration. This can't be done programmatically,
and so here we use an IO nemesis on n3 instead.

Part of https://github.com/cockroachdb/cockroach/issues/79215.
Closes #81834.

Release note: None

tbg commented:
Note to self: here's almost-ready code that verifies that r-amp stays bounded which should be added to this roachtest before merge. But maybe I can also use @kvoli's cluster-stats stuff which internally is issuing similar queries.

```go
func TestFoo(t *testing.T) {
	client, err := promapi.NewClient(promapi.Config{
		Address: "http://tobias-overload-new-0004.roachprod.crdb.io:9090",
	})
	now := timeutil.Now()
	require.NoError(t, err)
	c := promv1.NewAPI(client)
	// NB: this is, at each datapoint, the larger of the two values.
	const q = `
storage_l0_sublevels{store="3"}/20 > storage_l0_num_files{store="3"}/1000 or storage_l0_num_files{store="3"}/1000
`
	v, warns, err := c.QueryRange(context.Background(), q, promv1.Range{
		Start: now.Add(-10 * time.Minute),
		End:   now,
		Step:  time.Minute,
	})
	require.NoError(t, err)
	require.Len(t, warns, 0)

	m := v.(model.Matrix)

	var n int
	for _, ss := range m {
		for _, v := range ss.Values {
			n++
			// We're pausing once we hit .8, but there's some delay so we may overshoot a bit.
			// In practice, we're hoping to stay away from 1.0 most of the time and definitely
			// we shouldn't be hitting 2.0 (unless pausing is broken, in which case we're
			// sure to hit it).
			if v.Value > 2.0 {
				t.Errorf("%s at %s: overload score %.2f", ss.Metric, v.Timestamp, v.Value)
			}
		}
	}

	require.NotZero(t, n)

	if t.Failed() {
		t.Log(m)
	}
}
```

tbg commented:
Verified that this test still works with the latest updates to pausing, and in particular https://github.com/cockroachdb/cockroach/pull/85739.

<img width="822" alt="image" src="https://user-images.githubusercontent.com/5076964/184337082-d23496c4-89f8-46bf-9633-67a452cc0757.png">

tbg commented on pkg/cmd/roachtest/tests/admission_control_follower_overload.go:
```diff
@@ -0,0 +1,286 @@
+		//   └─md0         9:0    0 872.3G  0 raid0 /mnt/data1
+		//
+		// and so the actual write throttle is about 2x what was set.
+		c.Run(ctx, c.Node(3), "sudo", "systemctl", "set-property", "cockroach", "'IOWriteBandwidthMax={store-dir} 20971520'")
```
@data-matt

tbg replied:
@pavelkalinnikov
